{% extends 'TriplexDB/base.html' %}
{% load static %}

<head>
	<title>Gene detail view</title>
	<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" />
</head>
<style>
	.container {
		text-align: center;
		width: 100%;
	}

	.container img {
		display: block;
		margin: 0 auto;
	}

	figure {
		display: inline-block;
		/* Ensure the figure element is as wide as its content */
	}

	figcaption {
		text-align: center;
		width: 100%;
	}
</style>
{% block content %}
{% if object %}
<div class="table-responsive mx-5">
	{% if mouse %}
	<h4>Results for predicted triplexes involving transcripts of mouse Gene {{ gene }}</h4>
	<br />
	<h4>Mouse RNA Gene {{ gene }} is predicted to be involved in the formation of {{ nr_triplexes }} triplexes.</h4>
	{% else %}
	<h4>Results for predicted triplexes involving transcripts of human Gene {{ gene }}</h4>
	<br />
	<h4> Human RNA Gene {{ gene }} is predicted to be involved in the formation of {{ nr_triplexes }} triplexes.</h4>
	{% endif %}
	<br />
	<table class="table table-bordered table-hover table-lg display table-striped" id="results_table_rna"
		style="text-align: left; width: 100%">
		<thead class="table-dark" style="background-color: red;">
			<tr>
				<th scope="col">Gene ID</th>
				<th scope="col">Gene Symbol</th>
				<th scope="col">Transcript ID</th>
				<th scope="col">Transcript Chrom</th>
				<th scope="col">Transcript Start</th>
				<th scope="col">Transcript End</th>
				<th scope="col">Transcript Length</th>
				<th scope="col">Transcript Biotype</th>
				<th scope="col">Transcript Triplexcount</th>
				<th scope="col">Binding sites in UCSC</th>
			</tr>
		</thead>
		<tbody class="table-group-divider">
			{% for rna in object %}
			<tr>
				<td>{{rna.transcriptgeneid}}</td>
				<td>{{rna.transcriptgenesymbol}}</td>
				<td><a href="{% url 'transcript_detail' rna.transcriptid %}">{{rna.transcriptid}}</a></td>
				<td>{{ rna.transcriptchr }}</td>
				<td>{{ rna.transcriptstart }}</td>
				<td>{{ rna.transcriptend }}</td>
				<td>{{ rna.transcriptlength }}</td>
				<td>{{ rna.transcriptbiotype }}</td>
				<td><a href="{% url 'search_transcript_values_link' rna.transcriptid %}">
						{{rna.transcripttriplexcount}}
					</a></td>
				<td><a href="{{ ucsc_link }}" target="_blank" rel="noopener noreferrer">Binding sites in UCSC</a></td>
			</tr>
			{% endfor %}
		</tbody>
	</table>
</div>

<div class="container" style="display: flex; justify-content: space-between;">
	<div style="width: 48%;"> <!-- 48% width for each figure to make space between them -->
		<h5>Triplex forming region prediction for maximally scoring transcript</h5>
		<figure>
			<img src="{% static plot_path %}" style="width: 100%; height: auto;">
			<figcaption>Predicted triplex forming regions as represented by the summed negative logE values estimated by
				TriplexAligner are visualized along the length of the transcript in base pairs. Dark red indicates
				higher significance.</figcaption>
		</figure>
	</div>
	{% if circos_path %}
	<div style="width: 43%;"> <!-- 48% width for the second figure -->
		<h5>Circos plot</h5>
		<figure>
			<img src="{% static circos_path %}" style="width: 100%; height: auto;">
			<figcaption>Each line represents a predicted {{ transcript }} interaction with a DNA region on the
				corresponding chromosome.
				Line thickness scales with the negative logE value of the predicted interaction. </figcaption>
		</figure>
	</div>
	{% endif %}
</div>

{% else %}
<h3>No hits for this RNA in TripLexicon!</h3>
{% endif %}

{% endblock %}